KHR_XMP (#1553)
KhronosGroup/glTF: extensions/2.0/Khronos/KHR_xmp/README.md

# KHR_xmp

## Contributors

> TODO

Reviewer: We should add names here before merging.
@emilian0 
@robertlong 
@donmccurdy 
@andreasplesch 
@lexaknyazev 
@emackey 

Did I miss anyone?

| URL               | String formatted as described in the XMP documentation |
| Rational          | String formatted as described in the XMP documentation |
| FrameRate         | String formatted as described in the XMP documentation |
| FrameCount        | String formatted as described in the XMP documentation |
| Part              | String formatted as described in the XMP documentation |

### Additional Properties

| XMP Property Type | JSON Type |
| ----------------- | --------- |
| [ResourceEvent](https://github.com/adobe/xmp-docs/blob/master/XMPNamespaces/XMPDataTypes/ResourceEvent.md)| Object, we refer to the XMP documentation for a description of its properties |

## The XMP gltf Namespace
The `gltf` namespace is used to extend `XMP` to include metadata of interest to the glTF working group.
The table below describes the properties in the `gltf` namespace.

| Property Name  | Desrcription | JSON Type    |
| -------------- | ------------ | ------------ |
| spdx           | A [SPDX](https://spdx.org/licenses/) license identifier   | string |

## Defining XMP Metadata
An indirection level is introduced to avoid replicating the same XMP metadata in multiple glTF nodes.

> XMP metadatas are defined within a dictionary property in the glTF scene description file, by adding an extensions property to the top-level glTF 2.0 object and defining a KHR_xmp property with a `packets` array inside it.

Reviewer: `KHR_xmp` should use `` quotes.

The following example defines a glTF scene with a sample XMP metadata.

```
"extensions": {
    "KHR_xmp" : {
        "packets" : [
            {
                "dc:contributor" : [ "Creator1Name", "Creator2Email@email.com", "Creator3Name<Email@email.com>"],
                "dc:coverage" : "Bay Area, California, United States",
                "dc:creator" : [ "CreatorName", "CreatorEmail@email.com"],
                "dc:date" : ["1997-07-16T19:20:30+01:00"],
                "dc:description" : { "en-us": "text"},
                "dc:format" : "model/gltf-binary",
                "dc:identifier" : "urn:stock-id:292930",
                "dc:language" : ["en"],
                "dc:publisher" : ["Company"],
                "dc:relation" : ["https://www.khronos.org/"],
                "dc:rights" : {"en-us": "BSD"},
                "dc:source" : "http://related_resource.org/derived_from_this.gltf",
                "dc:subject" : ["architecture"],
                "dc:title" : { "en-us" : "MyModel", "it-IT" : "Mio Modello"},
                "dc:type" : ["Physical Object"],
                "gltf:spdx" : "BSD-Protection",
            }
        ]
    }
}
```

## Instantiating XMP metadata

> Metadata can be instantiate in a node by defining the extensions.KHR_xmp property and, within that, an index into the `packets` array using the `packet` property.

Reviewer 2: ```suggestion
Metadata can be instantiated in a node by defining the `extensions.KHR_xmp` property and, within that, an index into the `packets` array using the `packet` property.
```

The following example shows a glTF Mesh instantiating the XMP metadata at index `0`.

```
{
    "meshes": [
      {
        "extensions" : {
            "KHR_xmp": {
                "packet" : 0,
            }
        },
        "primitives": [
          {
            "attributes": {
                "NORMAL": 23,
                "POSITION": 22,
                "TANGENT": 24,
                "TEXCOORD_0": 25
            },
            "indices": 21,
            "material": 3,
            "mode": 4
          }
        ]
      }
    ]
}
```

### Notes:

#### Overriding Rule
A glTF scene might reference resources already containing XMP metadata.

> A relevant example would be an image node referencing a `png` file with embedded XMP metadata.

Reviewer 2: ```suggestion
A relevant example would be an image node referencing a PNG file with embedded XMP metadata.
```

When this happens the XMP metadata specified in the glTF manifest takes precedence over the metadata embedded in the referenced resource.

#### Transferring and merging metadata
When a glTF object (for example a Mesh) is copied from a glTF file to another, the associated KHR_xmp metadata should be copied as well.

> Copying metadata requires adding the source object metadata packet to the `KHR_xmp.packets` extension of the copied glTF and reference it from the copied object `KHR_xmp.packet` extension.

Reviewer: The phrasing of this second line is a little confusing. I suggest changing it to this:
"Copying metadata requires adding the source object's metadata packet to the destination glTF's `KHR_xmp.packets` array."

In the following example two glTFs containing a mesh with KHR_xmp metadata are copied into a third glTF.

glTF containing the first mesh:

```
{
    "meshes": [
      {
        "extensions" : { "KHR_xmp": { "packet" : 0 } },
        ...
      }
    ],
    "extensions": { "KHR_xmp" : { "packets" : [
      {
        "dc:title" : { "en-us" : "My first mesh"},
        ...
      },
    ]}},
}
```

glTF containing the second mesh:

```
{
    "meshes": [
      {
        "extensions" : { "KHR_xmp": { "packet" : 0 } },
        ...
      }
    ],
    "extensions": { "KHR_xmp" : { "packets" : [
      {
        "dc:title" : { "en-us" : "My second mesh"},
        ...
      },
    ]}},
}
```

glTF containing copies of both meshes:

```
{
    "meshes": [
      {
        "extensions" : { "KHR_xmp": { "packet" : 0 } },
        ...
      },
      {
        "extensions" : { "KHR_xmp": { "packet" : 1 } },
        ...
      }
    ],
    "extensions": { "KHR_xmp" : { "packets" : [
      {
        "dc:title" : { "en-us" : "My first mesh"},
        ...
      },
      {
        "dc:title" : { "en-us" : "My second mesh"},
        ...
      },
    ]}},
}
```

> The [xmpMM](https://github.com/adobe/xmp-docs/blob/master/XMPNamespaces/xmpMM.md) namespaces introduces mechanisms such as Pantry-Ingredient to handle the case when a glTF file is generated via derivation or composition of 1 or more source documents.

Reviewer 2: ```suggestion
The [xmpMM](https://github.com/adobe/xmp-docs/blob/master/XMPNamespaces/xmpMM.md) namespaces introduces mechanisms such as Pantry-Ingredient to handle the case when a glTF file is generated via derivation or composition of one or more source documents.
```

The following example defines a glTF scene with a sample XMP metadata.

```
"extensions": {
    "KHR_xmp" : {
        "packets" : [
            {
                "dc:contributor" : [ "Creator1Name", "Creator2Email@email.com", "Creator3Name<Email@email.com>"],
                "dc:coverage" : "Bay Area, California, United States",
                "dc:creator" : [ "CreatorName", "CreatorEmail@email.com"],
                "dc:date" : ["1997-07-16T19:20:30+01:00"],
                "dc:description" : { "en-us": "text"},
                "dc:format" : "model/gltf-binary",
                "dc:identifier" : "urn:stock-id:292930",
                "dc:language" : ["en"],
                "dc:publisher" : ["Company"],
                "dc:relation" : ["https://www.khronos.org/"],
                "dc:rights" : {"en-us": "BSD"},
                "dc:source" : "http://related_resource.org/derived_from_this.gltf",
                "dc:subject" : ["architecture"],
                "dc:title" : { "en-us" : "MyModel", "it-IT" : "Mio Modello"},
                "dc:type" : ["Physical Object"],

>                 "gltf:spdx" : "BSD-Protection",

Reviewer 3: We should probably use the term "license" instead of the term "spdx", and simply state that the value must be a valid SPDX license identifier. SPDX means "Software Package Data Exchange" and includes some things beyond licenses: https://spdx.org/about.

Author: 👍 Thanks Don, I think this is good feedback for Adam and the e-commerce WG once they will move spdx into their namespace.

* Ed Mackey, Analytical Graphics, Inc.

## Status

Draft

## Dependencies

Written against the glTF 2.0 spec.

## Overview

This extension adds support for [XMP (Extensible Metadata Platform)](https://github.com/adobe/xmp-docs) metadata to glTF.
XMP is a technology for embedding metadata into documents and has been [an ISO standard since 2012](https://www.iso.org/news/2012/03/Ref1525.html).
XMP metadata is embedded in a top-level glTF extension as an array of metadata packets.
XMP metadata packets can then be referenced from glTF objects of type: `asset`, `scene`, `node`, `mesh`, `material`, `image`, `animation`.
XMP metadata referenced by the glTF top level object `asset` applies to the entire glTF asset.
XMP metadata is organized in namespaces. This extension enables any XMP metadata namespace to be embedded in a glTF asset.
We refer to the [XMP namespaces documentation](https://github.com/adobe/xmp-docs/tree/master/XMPNamespaces) for a detailed description of XMP properties.

## XMP data types

This section describes how [XMP data types](https://github.com/adobe/xmp-docs/tree/master/XMPNamespaces/XMPDataTypes) shall be encoded in a glTF JSON.

> XMP supports three classes of data types: simple, structure and array. XMP structures are encoded via JSON objects. XMP arrays are encoded via JSON arrays.

Reviewer 3: Note that there are no XMP arrays present in the core properties below, and therefore no JSON arrays defined by this extension. Is this correct? Surely some XMP namespaces use arrays? I may be missing something.

Reviewer 3: Never mind, I think I see. Data types (e.g. arrays) may contain objects of the "core properties" types below, but are not themselves core properties.

### XMP Core Properties

The following table describes how [XMP Core Properties](https://github.com/adobe/xmp-docs/blob/master/XMPNamespaces/XMPDataTypes/CoreProperties.md) are encoded in the glTF manifest.

| XMP Property Type | JSON Type |
| ----------------- | --------- |
| [Boolean](https://github.com/adobe/xmp-docs/blob/master/XMPNamespaces/XMPDataTypes/CoreProperties.md#boolean)              | Boolean   |
| [Date](https://github.com/adobe/xmp-docs/blob/master/XMPNamespaces/XMPDataTypes/CoreProperties.md#date)                    | String formatted as described in the XMP documentation |
| [Integer](https://github.com/adobe/xmp-docs/blob/master/XMPNamespaces/XMPDataTypes/CoreProperties.md#integer)              | Number    |
| [Real](https://github.com/adobe/xmp-docs/blob/master/XMPNamespaces/XMPDataTypes/CoreProperties.md#real)                    | Number    |
| [Text](https://github.com/adobe/xmp-docs/blob/master/XMPNamespaces/XMPDataTypes/CoreProperties.md#text)                    | String |
| [Agent Name](https://github.com/adobe/xmp-docs/blob/master/XMPNamespaces/XMPDataTypes/CoreProperties.md#agent-name)        | String formatted as described in the XMP documentation |

> | [Choice](https://github.com/adobe/xmp-docs/blob/master/XMPNamespaces/XMPDataTypes/CoreProperties.md#choice)                | String    |

Reviewer 3: Is a Choice really just a string? I'm not sure what to do with this from the given link.

Author: that is a good point I just changed it to `A value chosen from a vocabulary of values of a specific type. `

> | [Language Alternative](https://github.com/adobe/xmp-docs/blob/master/XMPNamespaces/XMPDataTypes/CoreProperties.md#language-alternative) | A language alternative is a dictionary that maps a language code, specified via a lower-case [BCP-47](https://tools.ietf.org/html/bcp47) string, to text. See property `dc:title` in the example below|

Reviewer 3: The description at the link states:

> A default value, if known, should be the first array item. The order of other array items is not specified by this document.

We will lose that order by using a JSON dictionary here. Is that important for conversion to/from XML?

# KHR_xmp

## Contributors

* Emiliano Gambaretto, Adobe, <mailto:emiliano@adobe.com>
* Mike Bond, Adobe, [@MiiBond](https://twitter.com/MiiBond)
* Robert Long, Mozilla, [@arobertlong](https://twitter.com/arobertlong)
* Don McCurdy, Google, [@donmccurdy](https://twitter.com/donrmccurdy)
* Andreas Plesch, Individual Contributor,[@andreasplesch](https://github.com/andreasplesch)
* Alexey Knyazev, Individual Contributor, [@lexaknyazev](https://github.com/lexaknyazev)
* Ed Mackey, Analytical Graphics, Inc.

## Status

Draft

## Dependencies

Written against the glTF 2.0 spec.

## Overview

This extension adds support for [XMP (Extensible Metadata Platform)](https://github.com/adobe/xmp-docs) metadata to glTF.
XMP is a technology for embedding metadata into documents and has been [an ISO standard since 2012](https://www.iso.org/news/2012/03/Ref1525.html).
XMP metadata is embedded in a top-level glTF extension as an array of metadata packets.
Any JSON object in the glTF can reference metadata packets with the exception of the root object.
XMP metadata referenced by the glTF top level property `asset` applies to the entire glTF file.
XMP metadata is organized in namespaces.

> We refer to the [XMP namespaces documentation](https://github.com/adobe/xmp-docs/tree/master/XMPNamespaces) for a detailed description of XMP properties.

Reviewer 4: I would certainly hope that any/all namespaces are allowed - that's the whole point of having an *extensible* metadata standard

> We refer to the [XMP namespaces documentation](https://github.com/adobe/xmp-docs/tree/master/XMPNamespaces) for a detailed description of XMP properties.

Author: Yes, this extension doesn't make restrictions, it is just a way to transfer XMP metadata via a glTF.
If you are ok with this (we can bring this topic up during the next call) I am happy to make the change.

> We refer to the [XMP namespaces documentation](https://github.com/adobe/xmp-docs/tree/master/XMPNamespaces) for a detailed description of XMP properties.

Reviewer 3: Understood. I think the PR should say this explicitly, then.

### Additional Properties

| XMP Property Type | JSON Type |
| ----------------- | --------- |
| [ResourceEvent](https://github.com/adobe/xmp-docs/blob/master/XMPNamespaces/XMPDataTypes/ResourceEvent.md)| Object, we refer to the XMP documentation for a description of its properties |

## Defining XMP Metadata

An indirection level is introduced to avoid replicating the same XMP metadata in multiple glTF nodes.
XMP metadatas are defined within a dictionary property in the glTF scene description file by adding an `extensions` property to the top-level glTF 2.0 object and defining a `KHR_xmp` object. The `KHR_xmp` object defines two properties:

* `namespaces` : a dictionary mapping XMP [namespaces](https://github.com/adobe/xmp-docs/blob/master/Namespaces.md) to the URI where they are defined.
* `packets`: an array of metadata packets referencing the namespaces defined in the `namespaces` dictionary.

The following example defines a glTF scene with a sample XMP metadata.

```json
"extensions": {
    "KHR_xmp" : {
        "namespaces" : {

>             "dc" : "http://purl.org/dc/elements/1.1/"

Reviewer 4: Here is what I wrote to @emilian0 about this:
>The answer depends on whether the person is processing as JSON or JSON-LD.  If they are doing only JSON processing, then they can just look for dc:title and ignore the actual values of in @context.  However, if they are doing JSON-LD processing, then they will use the @context accordingly.   And again, this is just JSON/JSON-LD - nothing specific to XMP.

He pointed out that this PR is currently using `namespaces` instead of a proper `@context` to comply with ISO 16684-3.  So that needs to change as well to make the above more correct.

>             "dc" : "http://purl.org/dc/elements/1.1/"

Author: yes. Following Leonard recommendation I will make a pull request to move from our custom namespace mechanism to JSON-LD (that is used in ISO 16684-3). And yes, as a consequence, to correctly process the JSON-LD you will have to interpret the `@context` attribute.

>             "dc" : "http://purl.org/dc/elements/1.1/"

Reviewer 3: I'm not immediately managing to find a copy of `ISO 16684-3`, is that available somewhere? 

And, is there a reason for the `@` in `@context`? This would be the only occurrence of that prefix in the glTF specification.

>             "dc" : "http://purl.org/dc/elements/1.1/"

Reviewer 4: No, 16684-3 is still in development (DIS stage, in ISO speak).  

`@context` is required because the serialization is actually JSON-LD and not plain JSON.  JSON-LD requires the use of `@context` to define the namespaces used - see https://w3c.github.io/json-ld-syntax/#the-context

>             "dc" : "http://purl.org/dc/elements/1.1/"

Reviewer 3: I see, thanks. Should this specification mention JSON-LD somewhere? Will this information be useful to implementors of the extension?

>             "dc" : "http://purl.org/dc/elements/1.1/"

Reviewer 3: I'm frankly not familiar with JSON-LD or XMP, and I expect most implementors will not be either. If this is a significant motivation for the serialization/deserialization design I believe the topic should be given some explanation in the extension.

>             "dc" : "http://purl.org/dc/elements/1.1/"

Reviewer 4: Yes, JSON-LD should indeed be mentioned in the docs.  Good point...

>             "dc" : "http://purl.org/dc/elements/1.1/"

Reviewer 3: Bumping this back up — could JSON-LD be mentioned in this extension's documentation?

        },
        "packets" : [
            {
                "dc:contributor" : [ "Creator1Name", "Creator2Email@email.com", "Creator3Name<Email@email.com>"],
                "dc:coverage" : "Bay Area, California, United States",
                "dc:creator" : [ "CreatorName", "CreatorEmail@email.com"],
                "dc:date" : ["1997-07-16T19:20:30+01:00"],
                "dc:description" : { "en-us": "text"},
                "dc:format" : "model/gltf-binary",
                "dc:identifier" : "urn:stock-id:292930",
                "dc:language" : ["en"],
                "dc:publisher" : ["Company"],
                "dc:relation" : ["https://www.khronos.org/"],

>                 "dc:rights" : {"en-us": "BSD"},

Author: I believe they aren't part of RFC 3066, so I think the answer is not supported. I will ask Leonard for confirmation.

>                 "dc:rights" : {"en-us": "BSD"},

Reviewer 4: In the JSON-LD serialization of XMP (ISO 16684-3), you use `i-default` instead of `x-default` since (as noted in an earlier comment) we now use [BCP-47](https://tools.ietf.org/html/bcp47) codes and so that's `i-default`

>                 "dc:rights" : {"en-us": "BSD"},

Reviewer 3: I'm not sure what this implies for the `KHR_xmp` extension, sorry. As an implementor, where would I find this information? 😕

                "dc:source" : "http://related_resource.org/derived_from_this.gltf",
                "dc:subject" : ["architecture"],

>                 "dc:title" : { "en-us" : "MyModel", "it-IT" : "Mio Modello"},

Author: Thanks for looking into this Don. Yes I think we should avoid cherry-picking.
If you think it would beneficial to the glTF community we could specify that a KHR_xmp processor has to enforce lower-case for language alternative keys. I don't see any drawback to it, what do you think?

Do you think the current wording is clear enough `A language alternative is a dictionary mapping a language (RFC 3066 language code) to text` or should we perhaps reference the xmp specification you posted ?

Thanks

>                 "dc:title" : { "en-us" : "MyModel", "it-IT" : "Mio Modello"},

Reviewer 4: In the JSON-LD serialization of XMP (ISO 16684-3) which is what this work is based on, language codes are [BCP-47](https://tools.ietf.org/html/bcp47).  In section 2.1.1. of that standard, it is stated clearly:
>   At all times, language tags and their subtags, including private use
   and extensions, are to be treated as case insensitive: there exist
   conventions for the capitalization of some of the subtags, but these
   MUST NOT be taken to carry meaning.

>                 "dc:title" : { "en-us" : "MyModel", "it-IT" : "Mio Modello"},

Reviewer 3: Because JSON objects are case-sensitive, this means it wouldn't be safe for an implementation to do lookups on the `{ "en-us" : ... } ` key. The object must be converted to an array, and iterated over doing case-insensitive checks on the key. Perhaps this should be an array, if we are not going to make the case predictable.

There is also language in the XML spec to suggest that `x-default` must always be the _first_ item in the alternative array. I'm not sure I understand the reasoning for that, or whether it applies to us with a JSON serialization.

>                 "dc:title" : { "en-us" : "MyModel", "it-IT" : "Mio Modello"},

Reviewer 4: Good point @donmccurdy.  So make it a recommendation in your spec to make them lowercase.

It can't be an array since you need the object format to map the language to the value, in standard JSON.  In the full JSON-LD serialization, it uses the `@language` feature present there to handle this.

>                 "dc:title" : { "en-us" : "MyModel", "it-IT" : "Mio Modello"},

Reviewer 4: `x-default` isn't present in the JSON-LD serialization.

>                 "dc:title" : { "en-us" : "MyModel", "it-IT" : "Mio Modello"},

Reviewer 3: The [Core Properties](https://github.com/adobe/xmp-docs/blob/master/XMPNamespaces/XMPDataTypes/CoreProperties.md#language-alternative) language also says:

> A default value, if known, should be the first array item. The order of other array items is not specified by this document.

This requirement will be omitted from this extension, then,  and since this extension has no `@language` feature that information is lost. I don't have an opinion on whether that matters, but just want to confirm it.

>                 "dc:title" : { "en-us" : "MyModel", "it-IT" : "Mio Modello"},

Reviewer 4: @donmccurdy I am confused about what information you think is lost.  The language is simply done via the key instead of `@language`

                "dc:type" : ["Physical Object"],
            }
        ]
    }
}
```

## Instantiating XMP metadata

> Metadata can be instantiated in a node by defining the `extensions.KHR_xmp` property and, within that, an index into the `packets` array using the `packet` property.

Author: Hi Don yes I meant any JSON object in the schema. Does this sound reasonable? I can make the change (again we can bring this up during the next call).

> Metadata can be instantiated in a node by defining the `extensions.KHR_xmp` property and, within that, an index into the `packets` array using the `packet` property.

Reviewer 3: I'm not sure... let's discuss on the call.

The following example shows a glTF Mesh instantiating the XMP metadata at index `0`.

```json
{
    "meshes": [
      {
        "extensions" : {
            "KHR_xmp": {
                "packet" : 0,
            }
        },
        "primitives": [
          {
            "attributes": {
                "NORMAL": 23,
                "POSITION": 22,
                "TANGENT": 24,
                "TEXCOORD_0": 25
            },
            "indices": 21,
            "material": 3,
            "mode": 4
          }
        ]
      }
    ]
}
```

### Notes:

#### Overriding Rule

A glTF scene might reference resources already containing XMP metadata.
A relevant example would be an image node referencing a PNG file with embedded XMP metadata.

> When this happens the XMP metadata specified in the glTF manifest takes precedence over the metadata embedded in the referenced resource.

Reviewer 4: It's not about overriding but more of inheritance.  If someone sets a `dc:author` on a node, then they should be considered the author of everything on that node *that is not otherwise specified*.  So in the example you gave, if a given image/PNG had embedded XMP with a different `dc:author` that applies for that specific asset.

> When this happens the XMP metadata specified in the glTF manifest takes precedence over the metadata embedded in the referenced resource.

Reviewer 3: Your phrasing of precedence over everything "that is not otherwise specified" sounds like the exact opposite of the phrasing currently in this PR, "XMP metadata specified in the glTF manifest takes precedence over the metadata embedded in the referenced resource" to me — I think the PR should be reworded, if so?

> When this happens the XMP metadata specified in the glTF manifest takes precedence over the metadata embedded in the referenced resource.

Reviewer 4: Yes, the PR needs to be reworded.

> When this happens the XMP metadata specified in the glTF manifest takes precedence over the metadata embedded in the referenced resource.

Author: Taking your example Don, I was try to resolve the situation in which we have, for instance, an image object with KHR_xmp metadata that references a `png` with metadata. Which metadata should take precedence?
```
"images": [{
  "uri": "duckCM.png", 
  "extensions" : { "KHR_xmp": { "packet" : 0 } },
}]
```
My rationale was to give the option to the container (glTF) to override the referenced metadata.  I have no strong opinion though, I  just wanted to make sure this situation is addressed. It sounds like there is consensus in doing the opposite -> I am happy to make the change.

> When this happens the XMP metadata specified in the glTF manifest takes precedence over the metadata embedded in the referenced resource.

Reviewer 2: Supporting XMP metadata embedded in PNG files would require:
- parsing PNG chunks;
- decoding zlib/deflate compression;
- parsing XML-based XMP.

Do we expect web-based viewers to implement all of these?

> When this happens the XMP metadata specified in the glTF manifest takes precedence over the metadata embedded in the referenced resource.

Reviewer 4: Actually, it doesn't.  XMP is stored as plain-text in PNG (and pretty much every other file format), so you can do a simple regex (or equiv) search to find the XMP - but then yes, you'd need to parse the XML+RDF.

> When this happens the XMP metadata specified in the glTF manifest takes precedence over the metadata embedded in the referenced resource.

Reviewer 2: Ah, XMP specifically disallows deflate on its chunks, interesting. 

Still, would it be acceptable for glTF viewers to not implement embedded XMP support (PNG and JPEG) with this new spec update? So far, glTF web loaders just pass such images directly to the web browser without any processing.

> When this happens the XMP metadata specified in the glTF manifest takes precedence over the metadata embedded in the referenced resource.

Reviewer 4: I would think that is up to each processor @lexaknyazev , depending on what they are trying to achieve in their processing of the glTF

                "dc:subject" : ["architecture"],
                "dc:title" : { "en-us" : "MyModel", "it-it" : "Mio Modello"},
                "dc:type" : ["Physical Object"],
            }
        ]
    }
}
```

## Instantiating XMP metadata

Metadata can be instantiated in a gltf object by defining the `extensions.KHR_xmp` property and, within that, an index into the `packets` array using the `packet` property.
The following example shows a glTF Mesh instantiating the XMP metadata at index `0`.

```json
{
    "meshes": [
      {
        "extensions" : {
            "KHR_xmp": {
                "packet" : 0,
            }
        },
        "primitives": [
          {
            "attributes": {
                "NORMAL": 23,
                "POSITION": 22,
                "TANGENT": 24,
                "TEXCOORD_0": 25
            },
            "indices": 21,
            "material": 3,
            "mode": 4
          }
        ]
      }
    ]
}
```

### Notes:

#### Precedence

Metadata applied to JSON objects in the top level arrays ( `scenes`, `nodes`, `meshes`, `materials`, `images`, `animations`) has precedence over the metadata specified in the `asset` property of a glTF.

> A glTF might reference resources already containing XMP metadata. A relevant example would be an image object referencing a PNG/JPG file with embedded XMP metadata. The metadata embedded in the referenced resource takes precedence over the metadata specified in the glTF manifest. Note that glTF clients and viewers may ignore the metadata embedded in the referenced resources (for instance PNG/JPG images).

Reviewer 2: Aren't the last two statements contradicting each other?

Author: I don't think it is a direct contradiction but I agree this might be a source of confusion. 
This was the result of the conversation we both had with @donmccurdy . 
I believe the conclusion was that: 
* we want nested objects and referenced resources to have the right to override the metadata defined lower in the hierarchy.
* to free clients from having to parse XML-XMP in nested resources. 

The resolution was a note directed to glTF editors to make sure that all relevant metadata in the glTF is defined in the glTF manifest (so that client can ignore metadata in embedded resources).
Does this sound like a good compromise to you? 
This is the point of the last sentence, should it be more explicit? I can add `Therefore we recommend glTF editors to make sure all relevant metadata in the glTF is defined in the glTF manifest`

Reviewer 2: I got different conclusion from that conversation.
- Nested objects can override whatever they want as long as they keep everything in JSON.
- PNG/JPEG metadata is always ignored. Exporters should add metadata packets to `image` glTF objects if needed.

Author: Let me remove `The metadata embedded in the referenced resource takes precedence over the metadata specified in the glTF manifest.` I think it is indeed misleading.
`PNG/JPEG metadata is always ignored. Exporters should add metadata packets to image glTF objects if needed.` => yes this is what I was trying to say, I will make it more explicit.

Author: Hi @lexaknyazev, I believe I addressed your concerns, I am labelling this conversation resolved please double check and feel free to reopen/bring up additional concerns. Thank you.

                  {
                    "xmpMM:DocumentID" : "62bc2623-968e-4b09-8174-02dc53d6b856",
                    "xmpMM:InstanceID" : "1a33a91f-7351-471e-9b6c-24bca3213d2a",
                    "dc:title" : { "en-us" : "My first glTF"},
                    ...
                  },
                  {
                    "xmpMM:DocumentID" : "6d70a25e-129e-42d8-9d63-93bd9eb0298b",
                    "xmpMM:InstanceID" : "235b5571-e6a9-48fd-8d6b-b88276f37ee3",
                    "dc:title" : { "en-us" : "My second glTF"},
                    ...
                  }
                ],
                "xmpMM:Ingredients" : [
                  {
                    "xmpMM:DocumentID" : "62bc2623-968e-4b09-8174-02dc53d6b856",
                    "xmpMM:InstanceID" : "1a33a91f-7351-471e-9b6c-24bca3213d2a"
                  },
                  {
                    "xmpMM:DocumentID" : "6d70a25e-129e-42d8-9d63-93bd9eb0298b",
                    "xmpMM:InstanceID" : "235b5571-e6a9-48fd-8d6b-b88276f37ee3"
                  }
                ]
              }
            ]
        }
    }
}
```

### Implementation note

*This section is non-normative.*

> glTF viewers are free to chose which metadata in the glTF to display.
> Our reccomandation for viewers that choose to display XMP metadata is to include the metadata in the glTF `asset` object.
> 
> XMP is an extensible metadata format by design. In order to achieve better readibility and interoperability across the industry, we recommand glTF creators to use the following preferred list of namespace prefixes and URIs.

Reviewer 5: ```suggestion
glTF viewers are free to chose which metadata in the glTF to display.
Our recommendation for viewers that choose to display XMP metadata is to include the metadata in the glTF `asset` object.

XMP is an extensible metadata format by design. In order to achieve better readability and interoperability across the industry, we recommend glTF creators to use the following preferred list of namespace prefixes and URIs.
```

Author: Thanks
